### ISSUE TITLE
How to render / release directly to mp4 or gif, plus how to make a better gif with gifsky

### ISSUE THREAD

**Issue by arrowtype** [coldtype/coldtype#118]
I'm excited there is an easy way to render directly to mp4!

At first, I got stuck with the following error:

```
▶ coldtype sources/04-coldtype/bigger-loop.coldtype.py
    >>> watching... sources/04-coldtype/bigger-loop.coldtype.py
DEFAULT RENDER ACTION
> ffmpeg
writing mp4...
[image2 @ 0x7ff6d080b400] Could find no file with path '/Users/stephennixon/type-repos/shantell-sans/sources/04-coldtype/renders/bigger-loop.coldtype/scratch/scratch_%4d.png' and index in the range 0-4
/Users/stephennixon/type-repos/shantell-sans/sources/04-coldtype/renders/bigger-loop.coldtype/scratch/scratch_%4d.png: No such file or directory
...done
The file /Users/stephennixon/type-repos/shantell-sans/sources/04-coldtype/renders/bigger-loop.coldtype does not exist.
```

...even though I do have ffmpeg installed.

However, I eventually tried hitting `a` first to let the frames render, and _then_ hitting `r` to make an mp4, and... it works!

https://user-images.githubusercontent.com/45946693/183216667-ad7c20be-3203-4e93-8971-89e9d196e778.mp4

So, I guess my point of filing this issue is:
1. Documenting it in case some searches
2. Documenting it so that perhaps there could be a simpler error message, or better yet, perhaps a step to auto render frames if needed, then render the mp4, all with one `r` button press
3. Learn if I was doing something wrong! This is usually the case. 😅

**Comment by arrowtype** [comment]
Ah! Nice, and I've just realized how to make it release to gif:

```
release = my_animation_name.export("gif", loops=1)
```

**Comment by stenson** [comment]
Interesting! The `a` then `r` thing is definitely because, back when I was making lots of lyric videos, I was really only viewing the frames in an NLE, so the `r` step was all part of Premiere or something like that. Definitely makes sense that you'd want to do both though, and I'll admit I've shifted away from using NLE's for shorter videos as well. There is a keyboard shortcut (routed internally to `KeyboardShortcut.RenderAllAndRelease` in the keyboard.py source) that will do both a render & a release at once (the keyboard shortcut is shift+cmd+a), but that (like so much else) needs to be documented.

I'll also admit I don't regularly make gifs, so I haven't seen that funky color issue, though I did just replicate it on my setup. Pretty weird! I looked at the drawbot source and it looks like they use gifsicle internally. I tried that but had some trouble with it on the command line, though looks like gifski is another option, which does work. So one option for a better gif export would be to roll a gifski wrapper and call it from a custom release function, like this (would require a `brew install gifski` if you don't have it):

```python
from coldtype import *

@animation((600, 250), timeline=30, bg=hsl(0.7), render_bg=1)
def scratch(f):
    return (StSt("COLD", Font.ColdObvi(), 200
        , wdth=f.e("eeio"))
        .align(f.a.r)
        .f(1))

def gifski(a:animation, passes):
    from subprocess import run
    root = a.pass_path(f"%4d.{a.fmt}").parent.parent
    gif = root / (a.name + ".gif")
    run(["gifski", "--fps", str(scratch.timeline.fps), "-o", gif, *[p.output_path for p in passes if p.render == a]])

def release(passes):
    gifski(scratch, passes)
```

Will work on a better error message regardless!

**Comment by arrowtype** [comment]
Amazing, gifski seems to work perfectly! Thanks so much for including that code to make it so easy on my end.

**Comment by arrowtype** [comment]
Just as one extra note here, I found that gifsky makes GIFs at a max of about 800x600 by default, so to capture the default size of my composition, I had to add `"-W", str(a.rect[2])` to the call, like so:

```py
from coldtype import *

@animation((600, 250), timeline=30, bg=hsl(0.7), render_bg=1)
def scratch(f):
    return (StSt("COLD", Font.ColdObvi(), 200
        , wdth=f.e("eeio"))
        .align(f.a.r)
        .f(1))

def gifski(a:animation, passes):
    from subprocess import run
    root = a.pass_path(f"%4d.{a.fmt}").parent.parent
    gif = root / (a.name + ".gif")
    run(["gifski", "--fps", str(a.timeline.fps), "-o", gif, "-W", str(a.rect[2]), *[p.output_path for p in passes if p.render == a]])

def release(passes):
    gifski(scratch, passes)
```

**Comment by arrowtype** [comment]
Well, this solves my problem, so it's totally your call whether you'd want to close, or to rename if you want to take any actions from here. 

Thanks again!

**Comment by stenson** [comment]
ah nice note on `-W` — just added that to a little generic gifski wrapper made available in `coldtype.renderable.animation` (added an example to the repo here https://github.com/coldtype/coldtype/blob/main/examples/animations/transparent_gifski.py)

also added a better error message for the original issue, which will be released with 0.9.7 very shortly — will close for now, thanks for reporting!
